library/src/main/java/com/mux/video/upload/internal/TranscoderContext.kt

```diff
@@ -0,0 +1,592 @@
+    val MAX_ALLOWED_FRAMERATE = 120;
+    val MAX_ALLOWED_WIDTH = 1920
+    val MAX_ALLOWED_HEIGTH = 1080
+    val OPTIMAL_FRAMERATE = 30
+    val I_FRAME_INTERVAL = 5 // in seconds
+
+    private val extractor: MediaExtractor = MediaExtractor()
+    private var muxer: MediaMuxer? = null
+    private var videoTrackIndex = -10
+    private var audioTrackIndex = -10
+    private var outputVideoTrackIndex = -1
+    private var outputAudioTrackIndex = -1
+    // Used to configure decoders
+    private var inputAudioFormat: MediaFormat? = null
+    private var inputVideoFormat: MediaFormat? = null
+
+    // Used to configure encoders
+    private var outputAudioFormat: MediaFormat? = null
+    private var outputVideoFormat: MediaFormat? = null
+
+    // This is what decoder actually provide as an output, bit different then what we used to configure it
+    private var videoDecoderOutputFormat: MediaFormat? = null
+    private var decodedFrameWidth: Int = -1;
+    private var decodedFrameHeight: Int = -1;
+    private var targetedWidth = -1
+    private var targetedHeight = -1
+    private var targetedFramerate = -1
+    private var targetedBitrate = -1
+    private var scaledSizeYuv: Nv12Buffer? = null
+    val audioFrames = ArrayList<AVFrame>()
+
+    // Input parameters
+    private var inputWidth = -1
+    private var inputHeighth = -1
+    private var inputBitrate = -1
+    private var inputFramerate = -1
+
+    // Wait indefinetly for negative value, exit imidetly on 0, or timeout after a given us+
+    private var dequeueTimeout:Long = 0;
+    private var eofReached: Boolean = false;
+    private var transcodeAudio = false;
+    private var muxerConfigured = false;
+    private var numberOfDecodedFrames = 0;
+    private var numberOfEncodedFrames = 0;
+    private var numberOfInputFrames = -1;
+
+    private var videoDecoder:MediaCodec? = null
+    private var audioDecoder:MediaCodec? = null
+    private var videoEncoder:MediaCodec? = null
+    private var audioEncoder:MediaCodec? = null
+    public var  fileTranscoded = false
+    private var configured = false
+
+    init {
+        val cacheDir = File(appContext.cacheDir, "mux-upload")
+        cacheDir.mkdirs()
+        val destFile = File(cacheDir, UUID.randomUUID().toString() + ".mp4")
+        destFile.createNewFile()
+
+//        val cw = ContextWrapper(appContext)
+//        val directory = cw.getExternalFilesDir(Environment.DIRECTORY_DCIM)
+//        val testFile = File(directory, "output.mp4")
+//        val output = testFile.outputStream()
+        muxer = MediaMuxer(destFile.absolutePath, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
+        uploadInfo.standardizedFilePath = destFile.absolutePath
+//        muxer = MediaMuxer(testFile.absolutePath, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
+
+        try {
+            extractor.setDataSource(uploadInfo.file.absolutePath)
+            checkIfTranscodingIsNeeded()
+            configureDecoders()
```

Comment:
if `checkIfTranscodingIsNeeded()` results in false, would `configureDecoders()` still be needed?